[GTK][WPE][Skia] Introduce hybrid threaded CPU+GPU rendering mode

nikolaszimmermann opened the pull request:
#### 5ce158e9149ee257aab657a3d022fc29fdf25f9b
<pre>
[GTK][WPE][Skia] Introduce hybrid threaded CPU+GPU rendering mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=279618">https://bugs.webkit.org/show_bug.cgi?id=279618</a>

Reviewed by Carlos Garcia Campos.

Introduce a hybrid threaded rendering mode, scheduling tasks to both the
CPU and GPU worker pools, according to several possible strategies, that
can be selected using an environment variable. By default we use
CPU-affine rendering on WPE, and GPU-affine rendering on Gtk.

CPU-affine rendering:
1) If there is a non-identity device scale factor applied, dispatch to GPU pool
2) Dispatch to CPU pool if free workers are available
3) Dispatch to GPU pool if free workers are available
4) Randomly use either CPU or GPU pool, 50% each by default

GPU-affine rendering has 2) and 3) swapped.

Take into account the presence of accelerated ImageBuffer/NativeImage
references in the DisplayList. If we encounter them in a display list,
we can be sure that GPU rendering was used during recording. Therefore
in hybrid mode, we have to use a GPU worker thread for replay, which has
a GL/EGL context associated, and is able to wait for those fences that
signal completion of accelerated ImageBuffer/NativeImage rendering
completion.

Existing tests do not use hybrid mode by default -- otherwise lots of
0.01% pixel tests differences will appear, and non-determinism,
depending on CPU/GPU queue utilization -- we agreed to avoid testing
this for now.

* Source/WTF/wtf/WorkerPool.cpp:
(WTF::WorkerPool::numberOfTasks const):
Add accessor for the number of enqueued tasks.
* Source/WTF/wtf/WorkerPool.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/NativeImage.h:
(WebCore::NativeImageBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordResourceUse):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
(WebCore::DisplayList::RecorderImpl::usedAcceleratedRendering const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
(WebCore::SkiaPaintingEngine::SkiaPaintingEngine):
(WebCore::canPerformAcceleratedRendering):
(WebCore::SkiaPaintingEngine::recordDisplayList const):
(WebCore::SkiaPaintingEngine::isHybridMode const):
(WebCore::SkiaPaintingEngine::decideHybridRenderingMode const):
(WebCore::SkiaPaintingEngine::paintLayer):
(WebCore::SkiaPaintingEngine::minimumAreaForGPUPainting):
(WebCore::SkiaPaintingEngine::minimumFractionOfTasksUsingGPUPainting):
(WebCore::SkiaPaintingEngine::hybridPaintingStrategy):
(WebCore::SkiaPaintingEngine::renderingMode const): Deleted.
(WebCore::SkiaPaintingEngine::threadedRenderingMode const): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h:
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.setup_environ_for_server):
Disable hybrid mode for Gtk, only use GPU (WPE does the opposite).

Canonical link: <a href="https://commits.webkit.org/291106@main">https://commits.webkit.org/291106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc456280f4b8dfa395999ccc6856f5387c35ed93

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/855 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96733 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42419 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93827 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11680 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19777 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70462 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27957 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94778 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8922 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83149 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50789 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/91272 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8684 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/741 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41618 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84573 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78981 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/742 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98748 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90519 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18921 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13959 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79483 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19174 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78993 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78707 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19521 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23229 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/559 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11992 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18909 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24140 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113106 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18612 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32753 "Found 317 jsc stress test failures: jsc-layout-tests.yaml/js/script-tests/function-apply-many-args.js.layout-no-llint, microbenchmarks/array-from-object.js.lockdown, microbenchmarks/array-prototype-with-storage.js.bytecode-cache, microbenchmarks/array-prototype-with-storage.js.default, microbenchmarks/array-prototype-with-storage.js.dfg-eager ...") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22067 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20367 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3a7a4690c03c8b738603c5d6d41cd4a0d2ef92cc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/154415c324007dd21db49fa2697fecb83cc55755. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1b79b7bb4cabea63ce73246717a42dd534bbbc20. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9d1aaca82ac4412dc3223a4065f95714b4d710bf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -65,71 +52,42 @@ Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect
+
+    auto flags = Nicosia::Buffer::Flags { contentsOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha };
+    auto buffer = threadedPaintingPool->postPaintingTask(tileRect.size(), flags, [displayList = WTFMove(displayList), tileRect](Nicosia::Buffer& buffer) mutable {
         RELEASE_ASSERT(buffer.surface());
         if (auto* canvas = buffer.surface()->getCanvas()) {
```
If the buffer has a surface it should have a canvas too.

carlosgcampos replied:
Actually, I find beginThreadedGPUPainting a bit confusing, it's only needed for GPU and it doesn't have an endThreadedGPUPainting. What if we rename it as `canvas()` and make it pure virtual so that in case of non-accelerated buffer it just returns the can canvas and in the accelerated one it creates the buffer and returns the canvas? In that case we still need to check here that it's not nullptr.

nikolaszimmermann replied:
Nice idea, fixed.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+                m_pool.m_numberOfActiveCPUTasks++;
+        }
+
+        if (isAccelerated)
+            static_cast<Nicosia::AcceleratedBuffer*>(m_task.buffer.get())->beginThreadedGPUPainting();
```
I think it's a bit weird that the worker calls this.

nikolaszimmermann replied:
Agreed, will adapt (see the comment above).

carlosgcampos reviewed (commented):
Adding just a few comments for now. So, if I understood correctly, we are not limiting the amount of workers that can render with the GPU at the same time, right?

nikolaszimmermann commented:
> Adding just a few comments for now. So, if I understood correctly, we are not limiting the amount of workers that can render with the GPU at the same time, right?
No, because it is impossible to "limit" this properly - as we discussed on the channel last week.

We could keep track of the tasks that are currently executing a "GPU rendering task" - but these tasks are only doing the preparation operations for the actual GPU rendering, _on the CPU_. Once the GL commands will be submitted from Skia, the painting task is done for us -- but that doesn't mean the GPU even started to execute the actual GPU rendering.
And we do benefit from executing the preparation work in parallel, so that's why we didn't want to limit the amount of "currently executed tasks that will render to the GPU".

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -108,85 +132,98 @@ UnacceleratedBuffer::~UnacceleratedBuffer()
-        return m_painting.state == PaintingState::Complete;
+    m_fence = nullptr;
+
+    ensureOnMainThread([this, texture = WTFMove(m_texture)]() mutable {
+        PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
+        texture = nullptr;
```
Good catch -- this is in fact completly unnecessary, because the texture that's passed to AcceleratedBuffer, is always kept alive in the BitmapTexturePool owner by SkiaThreadedPaintingPool. So it should always have refCount==2 here. I think we should remove the main-thread destruction and rely on the BitmapTexturePool purging or reusing the texture later.

I will add an assertion to verify refCount == 2 here, ok?

nikolaszimmermann replied:
As expected, refCount is 2 here, no need to go to main thread.

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -108,85 +132,98 @@ UnacceleratedBuffer::~UnacceleratedBuffer()
 {
-    return { m_surface->width(), m_surface->height() };
-}
+    PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
```
Fixed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -108,85 +132,98 @@ UnacceleratedBuffer::~UnacceleratedBuffer()
+    externalTexture.fID = m_texture->id();
+    externalTexture.fFormat = GL_RGBA8;
+
+    auto size = m_texture->size();
```
Fixed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -108,85 +132,98 @@ UnacceleratedBuffer::~UnacceleratedBuffer()
+        SkColorSpace::MakeSRGB(),
+        &properties);
+
+    if (!m_surface || !m_surface->getCanvas())
+        return;
```
Good idea, fixed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+constexpr bool collectPaintTaskStatistics = false;
+constexpr Seconds dumpPaintTaskStatisticsInterval = 5_s;
+
+static unsigned long s_numberOfPaintTasks = 0;
+static unsigned long s_numberOfPaintTasksDispatchedToCPU = 0;
+static unsigned long s_numberOfPaintTasksDispatchedToGPU = 0;
+static unsigned long s_numberOfPaintTasksAboveGPUSizeThreshold = 0;
```
Fixed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+namespace WebCore {
+
+class SkiaThreadedPaintingPool::Worker : public AutomaticThread {
+protected:
```
This used to be the base-class for CPUWorker/GPUWorker classes, which I unified before posting the PR. Just a left-over.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+
+namespace WebCore {
+
+class SkiaThreadedPaintingPool::Worker : public AutomaticThread {
```
Agreed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -108,85 +132,98 @@ UnacceleratedBuffer::~UnacceleratedBuffer()
-        fence = nullptr;
-        surface = nullptr;
-    });
+    Buffer::beginPainting();
```
Correct, left-over from when this still contained more code. Fixed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+
+Ref<Nicosia::Buffer> SkiaThreadedPaintingPool::createAcceleratedBuffer(const IntSize& size, Nicosia::Buffer::Flags flags)
+{
+    PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
```
This doesn't work with Ref<Nicosia::Buffer>.

nikolaszimmermann replied:
But I can move the responsibility of makgin the GL context current to the caller:

```
    auto dispatchToGPU = [&](auto& locker) {
        if (!PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent())
            return dispatchToCPU(locker);

        if (collectPaintTaskStatistics)
            ++s_numberOfPaintTasksDispatchedToGPU;

        auto buffer = createAcceleratedBuffer(size, flags);
        buffer->beginPainting();
        m_tasks.append({ buffer.copyRef(), WTFMove(task) });
        m_condition->notifyOne(locker);
        return buffer;
    };
```

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.h:
```diff
@@ -1,4 +1,5 @@
 /*
+ * Copyright (C) 2017 Yusuke Suzuki <utatane.tea@gmail.com>.
```
The AutomaticThread logic, and parts of the code stem from WTF::WorkerPool, which was used before for the skiaUnacceleratedPaintingPool.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+
+    Locker locker { *m_lock };
+    for (unsigned i = m_workers.size(); i--;)
+        m_tasks.append({ nullptr, nullptr }); // Use null task to indicate that we want the thread to terminate.
```
Yes it does, fixed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+                m_pool.m_numberOfActiveCPUTasks--;
+        }
+
+        m_task.task = nullptr;
+        m_task.buffer = nullptr;
```
Fixed.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp:
```diff
@@ -0,0 +1,268 @@
+        bool isAccelerated = m_task.buffer->isBackedByOpenGL();
+
+        {
+            Locker locker { *m_pool.m_lock };
```
I can get rid of the locking in `work()` -- it was only used to track the number of active CPU/GPU tasks, which we no longer use in the scheduling logic.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5a90382295d6b730be17933576685be95c67df03. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/04c071f8cf40ad0dcddc3ef8f67a7b99f56b2b98. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

nikolaszimmermann commented:
The dependant PR #33568 landed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/619732fb609c2d171fc85e6d26fb7382a1f33679. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2fdd1a759c33d35a7fe6f6d31fd9a17ac4e9ce4d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5c5fa8382d481c81b3691c10aaf8f5f95f10e97d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/13912ecff51386aaba65d8cb74c52db4f48d1b01. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ccb19741c4afe3c2df67b21c83bd12645635ed7a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/29dcb9a0fc5d6115ef12c51fe28a823016f98b4f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/933b6817e135d8da853b8311915423a8ffa12ec8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/15baab6e43446d1c52c2f52221f98bdae6405f47. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3667fd886c56d6f805e07d765064771ffe3df4bf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

nikolaszimmermann commented:
This is no longer in draft state -- patch is ready for review. Performance wise it's a clear benefit on rpi4/imx8, and also on desktop style machines using low-end GPUs (thus makes sense to  choose the new 'CPU-affine rendering' strategy as default for WPE, and the 'GPU-affine rendering' for Gtk, assuming in general that Gtk has more high-end GPU users, than WPE (which is typically used on embedded). All the assumptions can be tuned at runtime, if our defaults are not sufficient for a certain use-case.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -275,6 +350,72 @@ unsigned SkiaPaintingEngine::numberOfGPUPaintingThreads()
+
+    std::call_once(onceFlag, [] {
+#if PLATFORM(WPE)
+        strategy = HybridPaintingStrategy::GPUAffineRendering; // Saturate CPU, before using GPU.
```
```suggestion
        strategy = HybridPaintingStrategy::CPUAffineRendering; // Saturate CPU, before using GPU.
```

nikolaszimmermann replied:
Thanks, last minute typo :/

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -143,17 +148,86 @@ static bool canPerformAcceleratedRendering()
+    auto handleGPUAffineRendering = [&]() -> bool {
+        // If the GPU worker pool has unused workers, use them.
+        if (m_gpuWorkerPool && m_gpuWorkerPool->numberOfTasks() < numberOfGPUPaintingThreads())
+            return false;
```
Shouldn't this return true?

nikolaszimmermann replied:
yes, there's a mistake in the logic, I'll revise it.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e9d24d3c62857b0cafebde2b34a1b5d8ad925897. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

carlosgcampos commented on Source/WTF/wtf/WorkerPool.h:
```diff
@@ -49,6 +49,8 @@ class WorkerPool : public ThreadSafeRefCounted<WorkerPool> {
 
     ASCIILiteral name() const { return m_name; }
 
+    unsigned numberOfTasks() const;
```
```suggestion
    WTF_EXPORT_PRIVATE unsigned numberOfTasks() const;
```

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/096cee04a22898744c026b6b0f86269acfc1b737)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91115 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10658 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/152 "Build is in progress. Recent messages:") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93165 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11064 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18976 "Build is in progress. Recent messages:") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94116 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/11064 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/152 "Build is in progress. Recent messages:") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/11064 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/152 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41010 "Build is in progress. Recent messages:") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83935 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/11064 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/152 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98100 "Build is in progress. Recent messages:Running configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89884 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18317 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/18976 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/98100 "Build is in progress. Recent messages:Running configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18576 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/152 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/98100 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/152 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18320 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112453 "Build is in progress. Recent messages:Running configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18044 "Build is in progress. Recent messages:") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/112453 "Build is in progress. Recent messages:Running configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21505 "Build is in progress. Recent messages:Running configuration") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19821 "Build is in progress. Recent messages:") | | | </details>

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e4e7d60452b00539036f8c735ebc7e72d0e8106a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/851d378c48310401044067f066cd5cfec13b1482)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91117 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/154 "Build is in progress. Recent messages:") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93167 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11066 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18978 "Build is in progress. Recent messages:") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94118 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/11066 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/154 "Build is in progress. Recent messages:") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/11066 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/154 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41012 "Build is in progress. Recent messages:") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83937 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/11066 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/154 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98102 "Build is in progress. Recent messages:Running configuration") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18319 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/18978 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/98102 "Build is in progress. Recent messages:Running configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18578 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/154 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/98102 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/154 "Build is in progress. Recent messages:") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21507 "Build is in progress. Recent messages:") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | </details>

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1dd58e3e7c4a85df38a849e2b3c0b1e83bc14058. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/85743c5ab2c3a36c711d85d7dc6fedd11ea05319. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4e29248ca8408b58d758b78fb1287c01e07e975a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/33582

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/a23377e78bdf8083960ac21314961a62e7aa9479)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91543 "Build is in progress. Recent messages:") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/606 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96512 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93593 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11452 "Build is in progress. Recent messages:") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/96512 "Build is in progress. Recent messages:Running configuration") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94544 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/11452 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/606 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/96512 "Build is in progress. Recent messages:Running configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/91038 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/11452 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/606 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41399 "Build is in progress. Recent messages:") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84345 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/11452 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/606 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98516 "Build is in progress. Recent messages:Running configuration") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90291 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/98516 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/606 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/98516 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/606 "Build is in progress. Recent messages:") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18700 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23976 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112875 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18411 "Build is in progress. Recent messages:") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/112875 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21871 "Build is in progress. Recent messages:Running configuration") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20177 "Build is in progress. Recent messages:") | | | </details>

nikolaszimmermann committed:
[GTK][WPE][Skia] Introduce hybrid threaded CPU+GPU rendering mode
https://bugs.webkit.org/show_bug.cgi?id=279618

Reviewed by Carlos Garcia Campos.

Introduce a hybrid threaded rendering mode, scheduling tasks to both the
CPU and GPU worker pools, according to several possible strategies, that
can be selected using an environment variable. By default we use
CPU-affine rendering on WPE, and GPU-affine rendering on Gtk.

CPU-affine rendering:
1) If there is a non-identity device scale factor applied, dispatch to GPU pool
2) Dispatch to CPU pool if free workers are available
3) Dispatch to GPU pool if free workers are available
4) Randomly use either CPU or GPU pool, 50% each by default

GPU-affine rendering has 2) and 3) swapped.

Take into account the presence of accelerated ImageBuffer/NativeImage
references in the DisplayList. If we encounter them in a display list,
we can be sure that GPU rendering was used during recording. Therefore
in hybrid mode, we have to use a GPU worker thread for replay, which has
a GL/EGL context associated, and is able to wait for those fences that
signal completion of accelerated ImageBuffer/NativeImage rendering
completion.

Existing tests do not use hybrid mode by default -- otherwise lots of
0.01% pixel tests differences will appear, and non-determinism,
depending on CPU/GPU queue utilization -- we agreed to avoid testing
this for now.

* Source/WTF/wtf/WorkerPool.cpp:
(WTF::WorkerPool::numberOfTasks const):
Add accessor for the number of enqueued tasks.
* Source/WTF/wtf/WorkerPool.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/NativeImage.h:
(WebCore::NativeImageBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordResourceUse):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
(WebCore::DisplayList::RecorderImpl::usedAcceleratedRendering const):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::finishAcceleratedRenderingAndCreateFence):
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
(WebCore::SkiaPaintingEngine::SkiaPaintingEngine):
(WebCore::canPerformAcceleratedRendering):
(WebCore::SkiaPaintingEngine::recordDisplayList const):
(WebCore::SkiaPaintingEngine::isHybridMode const):
(WebCore::SkiaPaintingEngine::decideHybridRenderingMode const):
(WebCore::SkiaPaintingEngine::paintLayer):
(WebCore::SkiaPaintingEngine::minimumAreaForGPUPainting):
(WebCore::SkiaPaintingEngine::minimumFractionOfTasksUsingGPUPainting):
(WebCore::SkiaPaintingEngine::hybridPaintingStrategy):
(WebCore::SkiaPaintingEngine::renderingMode const): Deleted.
(WebCore::SkiaPaintingEngine::threadedRenderingMode const): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h:
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.setup_environ_for_server):
Disable hybrid mode for Gtk, only use GPU (WPE does the opposite).

Canonical link: https://commits.webkit.org/291106@main

webkit-commit-queue commented:
Committed 291106@main (5ce158e9149e): <https://commits.webkit.org/291106@main>

Reviewed commits have been landed. Closing PR #33582 and removing active labels.
